extract.py

```diff
@@ -4,34 +4,114 @@
 """
 
 import os
+import re
 import igraph
 
 from igraph import Graph
 
-from clang.cindex import Index, CursorKind
+from clang.cindex import Index, CursorKind, TypeKind
 from clang import cindex
 
 from collections import Counter, deque
 from itertools import chain
 
+"""
+List all the headers from the folder.
+And formats them to be inserted in the newly generated c files.
+[ Works with lua only ]
+"""
+def get_headers(path): 
+    headers = []
+
+    for file in os.listdir(path):
+        if file.endswith(".h"):
+            if file == "ljumptab.h":
+                continue
+            if file == "ltests.h":
+                continue
+            headers.append(file)
+            
+    includes = """
+#include <time.h>
+#include <setjmp.h>
+#include <ctype.h>
+"""
+
+    for header in headers:
+        includes += "#include \"" + header + "\"\n"
+
+    return includes    
+
+"""
+Use the libclang extents to read from a file opend with 'r'
+"""
+def file_read_extent(file, extent):
+    file.seek(extent.start.offset)
+    return file.read(extent.end.offset - extent.start.offset)
+"""
+Common printing information about a decl.
+"""
+def print_decl(decl):
+    print(decl.displayname, decl.extent.start.line)
+
+
+"""
+Get the type declaration from things taht look like variables.
+It works on variables and function parameters.
+"""
+def variable_to_type(variable):
+    if variable.type.kind == TypeKind.POINTER:
+        decl = variable.type.get_pointee().get_declaration()
+        if decl.type.kind == TypeKind.TYPEDEF:
+            if decl.extent.start.line == 0:
+                print(variable.displayname, variable.extent.start.line)
+            return decl
+        else:
+            return None
+    elif variable.type.kind == TypeKind.TYPEDEF:
+        decl = variable.type.get_declaration()
+        if decl.extent.start.line == 0:
+            print(variable.displayname, variable.extent.start.line)
+        return decl
+    elif variable.type.kind == TypeKind.RECORD:
+        raise NameError("Error: " + str(variable.extent.start.line) + " " + variable )
```

Comment:
I think `NameError` is meant to mean "you referenced a variable `x` which hasn't been assigned to yet" in Python code itself. I think this is a user-level exception so should be `raise Exception`?

Reply:
I agree, I wanted to throw a different exception but the generic one should be sufficient.